bn-apps/ledger-sync/ledger-sync-service/src/main/kotlin/com/r3/businessnetworks/ledgersync/Helpers.kt

```diff
@@ -22,20 +40,52 @@ import sun.security.util.ByteArrayLexOrder
  * Note that this can be a lengthy list and no precautions are taken to ensure the output does not exceed the maximum
  * message size.
  */
-fun VaultService.withParticipants(vararg parties: Party): List<SecureHash> = queryBy<ContractState>(
-        VaultQueryCriteria(status = ALL),
-        PageSpecification(1, MAX_PAGE_SIZE)
-).states.filter {
-    it.state.data.participants.containsAll(parties.toList())
-}.map {
-    it.ref.txhash
+
+fun ServiceHub.withParticipants(vararg parties: Party, pageSize: Int = this.cordaService(ConfigurationService::class.java).pageSize()): List<SecureHash> {
+    val list = mutableListOf<SecureHash>()
+    val criteria = VaultQueryCriteria(status = ALL)
+    var count = 1
+    var page = vaultService.queryBy<ContractState>(criteria, PageSpecification(count, pageSize))
+    while(page.states.isNotEmpty()) {
+        page.states.filter {
+            it.state.data.participants.containsAll(parties.toList())
+        }.map {
+            list.add(it.ref.txhash)
+        }
+        count ++
+        page = vaultService.queryBy(criteria, PageSpecification(count, pageSize))
+    }
+    return list
 }
 
+
 /**
  * Calculates a compound hash of multiple hashes by hashing their concatenation in lexical order.
  */
 fun List<SecureHash>.hash(): SecureHash = map {
     it.bytes
 }.sortedWith(ByteArrayLexOrder()).fold(ByteArray(0)) { acc, hash ->
     acc + hash
-}.sha256()
\ No newline at end of file
+}.sha256()
+
+/*
+    read pageSize from ledgersync.conf. More config could be added if needed.
+ */
+@CordaService
+class ConfigurationService(appServiceHub : AppServiceHub): SingletonSerializeAsToken()  {
```

Comment:
Can you pls extend from the AbstractConfigurationService instead? Just to make sure that it's all consistent. Otherwise looks good to me.

Reply:
I cannot resolve AbstractConfigurationService and cannot find it in corda API neither... May I know where can I find it?